src/emulator/extensionsEmulator.ts

```diff
@@ -0,0 +1,141 @@
+import * as fs from "fs-extra";
+import * as os from "os";
+import * as path from "path";
+import { spawnSync } from "child_process";
+
+import * as planner from "../deploy/extensions/planner";
+import { FirebaseError } from "../error";
+import { toExtensionVersionRef } from "../extensions/refs";
+import { downloadExtensionVersion } from "./download";
+import { EmulatableBackend } from "./functionsEmulator";
+import { getExtensionFunctionInfo } from "../extensions/emulator/optionsHelper";
+
+export interface ExtensionEmulatorArgs {
+  projectId: string;
+  projectNumber: string;
+  aliases?: string[];
+  extensions: Record<string, string>;
+  projectDir: string;
+}
+// TODO: Consider a different name, since this does not implement the EmulatorInstance interface
```

Comment:
In theory, I wanted the emulated extensions to be just a type of EmulatableBackend and not another type of EmulatorInstance.

There are some problems with that ideal that you brought up here e.g. it's a lot nicer to have logging bulleted by "extensions" rather then "functions" when performing extensions related operations, but as you noted in the comments, we end up in a confusing state where ExtensionsEmulator is not really an emulator type because it relies heavily on the FunctionsEmulator for runtime things.

I think the alternative here is to implement "runtime delegate for extensions" (introduced to the emulator in https://github.com/firebase/firebase-tools/pull/4012) and have that be responsible for validating, building, and watching the source to load up in the emulator (https://github.com/firebase/firebase-tools/blob/d16a978df1be865a402a9713b29f75545b9384f4/src/deploy/functions/runtimes/index.ts#L53).

I'm certain there will be wrinkles in trying to implement that for the extensions emulator, but that might be an exercise that's worth doing (at least it was for the Functions side since now we use mostly same code for "deploying" functions to both the emulator and the production).

Happy to chat offline - I'm a bit cautious here since what I'm asking here is a large consideration and I don't know what other considerations you have in mind (e.g your launch schedule?)

Reply:
We should definitely chat offline on this, but I largely agree with your points here and think it's reasonable to refactor this. However, for launch planning purposes, I'd prefer to do so in a later PR (before merging to master). This would unblock the emulator UI work & let them test with something more realistic than an API spec. WDYT?

Reply:
please add me if you were to chat :)